lib/matplotlib/cm.py

```diff
@@ -235,6 +236,36 @@ def unregister_cmap(name):
     return _cmap_registry.pop(name)
 
 
+@functools.lru_cache(None)
+def _auto_norm_from_scale(scale_cls):
+    """
+    Automatically generate a norm class from *scale_cls*.
+
+    This differs from `.colors.make_norm_from_scale` in the following points:
+
+    - This function is not a class decorator, but directly returns a norm class
+      (as if decorating `.Normalize`).
+    - The scale is automatically constructed with ``nonpositive="mask"``, if it
+      supports such a parameter, to work around the difference in defaults
+      between standard scales (which use "clip") and norms (which use "mask").
+    - The returned norm class is memoized and reused for later calls.  (This
+      would be desirable in `.colors.make_norm_from_scale` as well, but
+      because that function sometimes gets called with `functools.partial`
+      instances which always compare unequal, a simple ``@lru_cache`` would be
+      insufficient.)
+    """
+    try:
+        norm = colors.make_norm_from_scale(
+            functools.partial(scale_cls, nonpositive="mask"))(
+            colors.Normalize)()
+    except TypeError:
+        norm = colors.make_norm_from_scale(scale_cls)(
+            colors.Normalize)()
+    norm_cls = type(norm)
+    norm_cls.__name__ = f"{scale_cls.__name__}Norm"
+    return norm_cls
```

Comment:
I guess this seems overkill to me, and fragile as shown by the pickling issue?  Why don't we just use `scale._scale_mapping` or make one, i.e. `colors._norm_mapping`?    Do we really need the extra flexibility given by allowing dynamic creation?

Reply:
I don't really want to create a separate mapping as people can register new scales (e.g. mpl-probscale does it), but they wouldn't show up in the norm mapping if these are separate.

Reply:
Back when I was a whippersnapper, we only had the linear norm and we liked it....

I'm not sure that making this super-flexible is worth the complication.  The strings can be for a few common norms.  How many specialized scales truly need a norm as well, and how much of a problem is it to ask users to just pass the norm instead to the string?  Its not like probscale is going to be a super-useful color norm.  If you wanted to add a `register_norm`, fine, but making it all dynamic just seems to be asking for trouble.

Reply:
#20755 fixes that part.